### ISSUE TITLE
Proposal: Model.rename() method

### ISSUE THREAD

**Issue by JedWatson** [Automattic/mongoose#3172]
Recent adventures renaming paths during schema migration have made me think this could be a nice addition to the `Model` API:

```
MyModel.rename({ oldPath: 'newPath' }, callback);
```

This is basically shorthand for:

```
MyModel.collection.update({}, { $rename: { oldPath: 'newPath' } }, { multi: true }, callback);
```

It would implement the same behaviour as `Model.update`, i.e. return a `Query` if no `callback` argument is provided, etc.

I've seen enough support requests for Keystone about how to handle schema migration that I'm considering adding support for this to the updates framework, am happy to submit a PR for mongoose if it would be welcome here instead.

**Comment by vkarpov15** [comment]
I'm hesitant to get into schema migrations because they really don't play nice with mongoose casting. If you where to rename a field, unless the new name of the field were already defined in the schema or you had `strict: false`, you wouldn't see that field when you loaded docs from the db. Also, that `update()` operation can be extremely expensive for large collections.

Long story short, I'd recommend implementing this as a plugin using [statics](http://mongoosejs.com/docs/guide.html#statics) if it'll be a good fit for keystone but you're interested in sharing it for people who think it would be useful.

**Comment by JedWatson** [comment]
Good point about update being expensive. My reasoning for adding this was mainly due to spending a fairly large amount of time researching how it should _actually_ be done, after a few questions on Keystone's support groups. Turns out it's supported in MongoDB itself, it's just not that easy to find through the layers of abstraction. tl;dr: read the MongoDB docs :)

re: the field being defined in the schema, the main use-case seems to be wanting to change a path in a schema at some point during development (presumably after the app is in production). It's easy to update the path in the Schema, not so intuitive to work out how to migrate the data across in your existing documents. Once the path changes the data seems to "disappear".

Will look into adding this as a plugin.

**Comment by vkarpov15** [comment]
Yeah migrations are something that mongoose doesn't really have a good answer for. Doing `update()` across the entire collection is easy, but it doesn't really work for large collections and I'm hesitant to make that mongoose's official mechanism for migrations if it's not going to work for large collections.

**Comment by djanowski** [comment]
Would it make sense to allow for aliases in fields? This would allow to rename a field in the model, deploy the new code, and run the migration, all without inconsistencies (both versions of the code can read the field, whatever the name in the database).

**Comment by vkarpov15** [comment]
@djanowski you can do that with plain old getters and setters (or virtuals) if I'm understanding your suggestion correctly...
